ambry-store/src/main/java/com.github.ambry.store/CompactionManager.java

```diff
@@ -47,9 +54,109 @@ CompactionDetails getCompactionDetails(BlobStore blobStore) throws StoreExceptio
+                    blobStoreStats.getErrorMargin());
+            // @TODO: introduce storeConfig for minimum benefit to consider
+            if (costBenefitInfo.benefit > 1 && (bestCandidateToCompact == null
+                || costBenefitInfo.costBenefitRatio < bestCandidateToCompact.costBenefitRatio)) {
+              bestCandidateToCompact = costBenefitInfo;
+              logger.trace("Updating best candidate to compact to " + bestCandidateToCompact);
+            }
+            endEntry = validDataPerLogSegments.getSecond().lowerEntry(endEntry.getKey());
+          }
+          firstEntry = validDataPerLogSegments.getSecond().higherEntry(firstEntry.getKey());
+        }
+        if (bestCandidateToCompact != null) {
+          List<String> segmentsToCompact = new ArrayList<>(validDataPerLogSegments.getSecond()
+              .subMap(bestCandidateToCompact.firstLogSegmentName, true, bestCandidateToCompact.lastLogSegmentName, true)
+              .keySet());
+          details = new CompactionDetails(validDataPerLogSegments.getFirst(), segmentsToCompact);
+        }
       }
     }
     return details;
   }
+
+  /**
+   * Calculates and returns the {@link CostBenefitInfo} for a given candidate
+   * @param firstLogSegmentName first log segment name of the candidate
+   * @param lastLogSegmentName last log segment name of the candidate
+   * @param validDataSizePerLogSegment Valid data size map
+   * @param segmentCapacity Segment Capacity of the {@link BlobStore}
+   * @param segmentHeaderSize Segment header size of the {@link BlobStore}
+   * @param errorMargin Error margin to factor in while calculating the cost benefit
+   * @return the {@link CostBenefitInfo} for the given candidate
+   */
+  private CostBenefitInfo getCostBenefitInfo(String firstLogSegmentName, String lastLogSegmentName,
+      NavigableMap<String, Long> validDataSizePerLogSegment, long segmentCapacity, long segmentHeaderSize,
+      long errorMargin) {
+    // cost = total valid data size (or data that needs to be copied over)
+    // benefit = no of segments reclaimed = (total source segments - total target segments)
+    //           total target segments = Ceil (total cost / (segment capacity - segment header - error margin))
+    long totalCost = 0;
+    int totalSegmentsToBeCompacted = 0;
+    String curSegmentName = firstLogSegmentName;
+    while (!curSegmentName.equals(lastLogSegmentName)) {
+      totalCost += validDataSizePerLogSegment.get(curSegmentName);
+      curSegmentName = validDataSizePerLogSegment.higherEntry(curSegmentName).getKey();
+      totalSegmentsToBeCompacted++;
+    }
+    // add the last log segment
+    totalCost += validDataSizePerLogSegment.get(curSegmentName);
+    totalSegmentsToBeCompacted++;
+    long maxCapacityPerSegment = segmentCapacity - segmentHeaderSize - errorMargin;
+    int benefit = totalSegmentsToBeCompacted - (int) Math.ceil(totalCost / (maxCapacityPerSegment * 1.0));
```

Comment:
yes, we want to find the actual double value of (cost / capacity) (which is long/ long) and then take ceil over it.